netty/src/test/java/io/grpc/netty/NettyStreamTestBase.java

```diff
@@ -95,14 +88,16 @@
   public void setUp() {
     MockitoAnnotations.initMocks(this);
 
-    mockFuture(true);
-    when(channel.write(any())).thenReturn(future);
-    when(channel.writeAndFlush(any())).thenReturn(future);
     when(channel.alloc()).thenReturn(UnpooledByteBufAllocator.DEFAULT);
     when(channel.pipeline()).thenReturn(pipeline);
     when(channel.eventLoop()).thenReturn(eventLoop);
     when(channel.newPromise()).thenReturn(new DefaultChannelPromise(channel));
     when(channel.voidPromise()).thenReturn(new DefaultChannelPromise(channel));
+    ChannelPromise completedPromise = new DefaultChannelPromise(channel)
+        .setSuccess();
+    when(channel.write(any())).thenReturn(completedPromise);
+    when(channel.writeAndFlush(any())).thenReturn(completedPromise);
+    when(writeQueue.enqueue(any(QueuedCommand.class), anyBoolean())).thenReturn(completedPromise);
```

Comment:
Instead of these stubbings, I would return a ChannelPromise instance without setting success. And set success only in test method that needs it (i.e. `notifiedOnReadyAfterWriteCompletes()`) 

```java
when(channel.write(any())).thenReturn(new DefaultChannelPromise(channel));
when(channel.writeAndFlush(any())).thenReturn(new DefaultChannelPromise(channel));
when(writeQueue.enqueue(any(QueuedCommand.class), anyBoolean()))
    .thenReturn(new DefaultChannelPromise(channel));
```

Here `writeQueue` is a mock, so returning a non-null object is only to avoid NPE. Whether it's success or done or failure is irrelevant in most of the tests. Some tests may require a definite state of the returned future and it should be set explicitly in those tests.

Reply:
I don't want tests that require or implicitly depend on the state of every mock in the system. Only the tests that need the future to be delayed should be aware of the future mocking. Having never-completed futures is simply a bad idea.